pkg/issuer/acme/http/istio.go

```diff
@@ -0,0 +1,201 @@
+/*
+Copyright 2021 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package http
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	cmacme "github.com/jetstack/cert-manager/pkg/apis/acme/v1"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+	"github.com/jetstack/cert-manager/pkg/util"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+func newUnstructuredVirtualService(
+	group string,
+	version string,
+	kind string,
+	generateName string,
+	namespace string,
+	labels map[string]string,
+	ownerReferences *[]metav1.OwnerReference,
+	spec *map[string]interface{},
+) *unstructured.Unstructured {
+	unstr := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": group + "/" + version,
+			"kind":       kind,
+			"spec":       *spec,
+		},
+	}
+
+	unstr.SetGenerateName(generateName)
+	unstr.SetNamespace(namespace)
+	unstr.SetLabels(labels)
+	unstr.SetOwnerReferences(*ownerReferences)
+
+	return unstr
+}
```

Comment:
Instead of dealing with the unstructured types, could we not create an internal type that we can work from? Similar to how istio.io/client-gen is doing things by wrapping the istio.io/api with the k8s fields: https://github.com/istio/client-go/blob/5ecafdc454fa37e919821491370738e909e20b3a/pkg/apis/networking/v1beta1/types.gen.go#L249

```go
import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	istiometav1alpha1 "istio.io/api/meta/v1alpha1"
	networkingv1beta1 "istio.io/api/networking/v1beta1"
)

type VirtualService struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec networkingv1beta1.VirtualService `json:"spec,omitempty"

	Status istiometav1alpha1.IstioStatus `json:"status"`
}
```

Reply:
What would be the best place in the project file structure for this file, also does this work with DeepCopy?